[dtensor] add grad placements kwarg to to_local API

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110629

When we convert to local tensor, dtensor can't track autograd or
gradient layout of the local tensor anymore, if user do sth not expected, there
needs to be a way for user to hint about the gradient layout of the
local tensor

wanchaol committed:
[dtensor] add grad placements kwarg to to_local API

When we convert to local tensor, dtensor can't track autograd or
gradient layout of the local tensor, if user do sth not expected, there
needs to be a way for user to hint about the gradient layout of the
local tensor

[ghstack-poisoned]

zdevito commented on torch/distributed/_tensor/api.py:
```diff
@@ -301,19 +317,33 @@ def from_local(
 
+        Keyword args:
+            grad_placements (List[:class:`Placement`], optional): the placements that
+                describes the gradient layout of the local torch.Tensor on DeviceMesh.
```
Maybe a clearer way to describe this, if I understand correctly: the placement describes the future layout of any gradient of the tensor returned from this function.

wanchaol replied:
sounds great, thanks updated!

wanchaol committed:
Update on "[dtensor] add grad placements kwarg to to_local API"


When we convert to local tensor, dtensor can't track autograd or
gradient layout of the local tensor anymore, if user do sth not expected, there
needs to be a way for user to hint about the gradient layout of the
local tensor

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add grad placements kwarg to to_local API"


When we convert to local tensor, dtensor can't track autograd or
gradient layout of the local tensor anymore, if user do sth not expected, there
needs to be a way for user to hint about the gradient layout of the
local tensor

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge
